src/backend/access/heap/heapam.c

```diff
@@ -400,19 +400,32 @@ heapgetpage(TableScanDesc sscan, BlockNumber page)
 	CHECK_FOR_INTERRUPTS();
 
 	/* Prefetch next block */
-	if (enable_seqscan_prefetch)
+	if (enable_seqscan_prefetch && seqscan_prefetch_buffers > 0)
 	{
-		int prefetch_limit = seqscan_prefetch_buffers;
+		uint32 prefetch_limit = seqscan_prefetch_buffers;
+		BlockNumber	prefetch_start = page;
 		ParallelBlockTableScanWorker pbscanwork = scan->rs_parallelworkerdata;
+
 		if (pbscanwork != NULL && pbscanwork->phsw_chunk_remaining < prefetch_limit)
 			prefetch_limit = pbscanwork->phsw_chunk_remaining;
-		if (page + prefetch_limit >= scan->rs_nblocks)
-			prefetch_limit = scan->rs_nblocks - page - 1;
+
+		/*
+		 * If we've started prefetching buffers, we don't have to prefetch
+		 * all pages in page..page + n, but only page + n; saving N-1 prefetch
+		 * calls.
```

Comment:
The phrasing sounds a bit weird. It's true that we save "save N-1 prefetch calls", but you wouldn't really expect this code to do that on every call anyway, it would be silly. (I know that's exactly what we are doing before this PR, but it is silly :-) )

```suggestion
		 * If this is the first page, initiate prefetch of pages page..page + n. On
		 * each subsequent call, prefetch the next page that we haven't prefetched
		 * yet, at page + n.
```